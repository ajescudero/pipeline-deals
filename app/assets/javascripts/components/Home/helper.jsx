export async function sortEntries(result) {
  return result.data.entries.sort(
    (a, b) => a["deal_stage"]["percent"] - b["deal_stage"]["percent"]
  );
}
export function getLabels(labels) {
  return _.map(_.uniqBy(labels, "deal_stage.name"), "deal_stage.name");
}

export function getDeals(labels, entries) {
  return labels.map((label) => {
    let sum = 0;

    for (let entry of entries.filter(
      (entry) => entry.deal_stage.name == label
    )) {
      sum += parseFloat(entry.value);
    }

    return _.round(sum, 2);
  });
}
