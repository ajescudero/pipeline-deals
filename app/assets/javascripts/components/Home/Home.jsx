import React, { useState, useEffect } from "react";
import axios from "axios";
import _ from "lodash";
import { Bar } from "react-chartjs-2";
import PropTypes from "prop-types";
import { sortEntries, getLabels, getDeals } from "./helper";

const API = `${process.env.REACT_APP_API_URL}/deals.json?api_key=${process.env.REACT_APP_API_KEY}`;

function Home() {
  const [dataSet, setDataSet] = useState({});
  const [options, setOptions] = useState({});
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    async function fetchData() {
      try {
        const result = await axios.get(API);
        const entries = await sortEntries(result);

        let labels = getLabels(entries);

        let deals = getDeals(labels, entries);

        const chartState = {
          labels: labels,
          datasets: [
            {
              backgroundColor: "rgba(75,192,192,1)",
              borderColor: "rgba(0,0,0,1)",
              borderWidth: 2,
              data: deals,
            },
          ],
        };

        const options = {
          title: {
            display: true,
            text: "Total $ value of deals per stage",
            fontSize: 20,
          },
          legend: {
            display: false,
          },
          scales: {
            yAxes: [
              {
                ticks: {
                  callback: function (label) {
                    return (
                      "$" +
                      label.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",")
                    );
                  },
                  beginAtZero: true,
                  fontSize: 10,
                },
                gridLines: {
                  display: true,
                },
              },
            ],
          },
          tooltips: {
            mode: "label",
            label: "mylabel",
            callbacks: {
              label: function (tooltipItem) {
                return (
                  "$" +
                  tooltipItem.yLabel
                    .toString()
                    .replace(/\B(?=(\d{3})+(?!\d))/g, ",")
                );
              },
            },
          },
        };

        setDataSet(chartState);
        setOptions(options);
        setIsLoading(false);
      } catch (error) {
        console.error(error);
      }
    }
    fetchData();
  }, []);

  return (
    <>{isLoading == false ? <Bar data={dataSet} options={options} /> : false}</>
  );
}

Home.propTypes = {
  dataSet: PropTypes.array,
  isLoading: PropTypes.bool,
};

export default Home;
