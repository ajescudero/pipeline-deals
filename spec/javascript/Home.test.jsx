import React from "react";
import Home from "../../app/assets/javascripts/components/Home/Home";
import { shallow, render, mount, invoke } from "enzyme";
import ReactLoading from "react-loading";
import { Bar } from "react-chartjs-2";
jest.mock("axios");

describe("Home", () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<Home debug />);

    expect(component).toMatchSnapshot();
  });

  describe("<ReactLoading />", () => {
    it("should render loading spinner", () => {
      const container = render(<ReactLoading />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe("fetchData", () => {
  test("successfully data from an API", () => {
    const wrapper = mount(<Bar />);
    wrapper.invoke("handler")();
    expect(wrapper.text()).toBe("Hi Yusinto Ngadiman React Junkie!");
  });
});
