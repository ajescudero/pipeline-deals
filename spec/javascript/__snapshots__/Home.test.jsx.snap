// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home <ReactLoading /> should render loading spinner 1`] = `undefined`;

exports[`Home should render correctly in "debug" mode 1`] = `
<Fragment>
  <styled.div>
    <div>
      <Loading
        color="grey"
        delay={0}
        height={64}
        type="bars"
        width={64}
      />
      Loading ...
    </div>
  </styled.div>
</Fragment>
`;
